pg_lakehouse/src/datafusion/schema.rs

```diff
@@ -0,0 +1,119 @@
+use async_std::sync::Mutex;
+use async_std::task;
+use async_trait::async_trait;
+use datafusion::catalog::schema::SchemaProvider;
+use datafusion::datasource::TableProvider;
+use datafusion::error::Result;
+use pgrx::*;
+use std::any::Any;
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::collections::HashMap;
+use std::sync::Arc;
+use supabase_wrappers::prelude::*;
+
+use crate::fdw::handler::*;
+use crate::fdw::options::*;
+use crate::schema::attribute::*;
+
+use super::catalog::CatalogError;
+use super::format::*;
+use super::provider::*;
+
+pub struct LakehouseSchemaProvider {
+    schema_name: String,
+    tables: Arc<Mutex<HashMap<pg_sys::Oid, Arc<dyn TableProvider>>>>,
+}
+
+impl LakehouseSchemaProvider {
+    pub fn new(schema_name: &str) -> Self {
+        Self {
+            schema_name: schema_name.to_string(),
+            tables: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    async fn table_impl(&self, table_name: &str) -> Result<Arc<dyn TableProvider>, CatalogError> {
+        let pg_relation = unsafe {
+            PgRelation::open_with_name(table_name).unwrap_or_else(|err| {
+                panic!("{:?}", err);
+            })
+        };
+
+        let mut tables = self.tables.lock().await;
+
+        let table = match tables.entry(pg_relation.oid()) {
+            Occupied(entry) => entry.into_mut(),
+            Vacant(entry) => {
+                let table_options = pg_relation.table_options()?;
+                let path = require_option(TableOption::Path.as_str(), &table_options)?;
+                let extension = require_option(TableOption::Extension.as_str(), &table_options)?;
+                let format = require_option_or(TableOption::Format.as_str(), &table_options, "");
+
+                let mut attribute_map: HashMap<usize, PgAttribute> = pg_relation
+                    .tuple_desc()
+                    .iter()
+                    .enumerate()
+                    .map(|(index, attribute)| {
+                        (
+                            index,
+                            PgAttribute::new(attribute.name(), attribute.atttypid),
+                        )
+                    })
+                    .collect();
+
+                let provider = match TableFormat::from(format) {
+                    TableFormat::None => create_listing_provider(path, extension).await?,
+                    TableFormat::Delta => create_delta_provider(path, extension).await?,
+                };
+
+                for (index, field) in provider.schema().fields().iter().enumerate() {
+                    if let Some(attribute) = attribute_map.remove(&index) {
+                        can_convert_to_attribute(field, attribute)?;
+                    }
+                }
+
+                entry.insert(provider)
+            }
+        };
+
+        Ok(table.clone())
+    }
+}
+
+#[async_trait]
```

Comment:
Do we definitely need async_trait? async traits have been stabilized, our latest Rust should allow it without the extra crate.

Reply:
This happens when we remove it:

```
error[E0195]: lifetime parameters or bounds on method `table` do not match the trait declaration
  --> pg_lakehouse/src/datafusion/schema.rs:96:19
   |
96 |     async fn table(&self, table_name: &str) -> Result<Option<Arc<dyn TableProvider>>> {
   |                   ^ lifetimes do not match method in trait
```

Reply:
When I remove `async_trait` and expand `table` to return a `Pin<Box<..>>` as per https://docs.rs/async-trait/latest/async_trait/, I get the error that ` the trait `std::marker::Send` is not implemented for `NonNull<RelationData>` which seems to come rom the fact that `PgRelation` doesn't implement `Send`. I don't understand why using `async_trait` gets rid of this error...if we can figure that out we can get rid of it probably. Also, I looked and DataFusion uses `async_trait` for its `SchemaProvider` traiit implementation.

Reply:
That makes sense, if one of our dependencies uses it then we're probably stuck with it.